### ISSUE TITLE
iD sometimes creates ways with duplicated nodes (way lenght = 0)

### ISSUE THREAD

**Issue by naoliv** [openstreetmap/iD#3211]
While fixing some geometries I found that sometimes people editing with iD are creating ways with duplicated nodes (ie, the initial and final nodes are the same)

For example, https://www.openstreetmap.org/api/0.6/way/427739762

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<osm version="0.6" generator="CGImap 0.4.3 (6182 thorn-05.openstreetmap.org)" copyright="OpenStreetMap and contributors" attribution="http://www.openstreetmap.org/copyright" license="http://opendatacommons.org/licenses/odbl/1-0/">
 <way id="427739762" visible="true" version="1" changeset="40334708" timestamp="2016-06-28T04:06:07Z" user="AjBelnuovo" uid="1799626">
  <nd ref="1840070954"/>
  <nd ref="1840070954"/>
  <tag k="highway" v="residential"/>
  <tag k="name" v="Rua Hermes Salgado Vasconcelos"/>
 </way>
</osm>
```

Is it possible to avoid this, please?

**Comment by pnorman** [comment]
> While fixing some geometries I found that sometimes people editing with iD are creating ways with duplicated nodes (ie, the initial and final nodes are the same)

What you describe is not an error by itself or necessarily a duplicated node, circular ways must have the same initial and final nodes.

This particular case is an error because it's a two node ways where both nodes are the same.

**Comment by bhousel** [comment]
^  That does give me a clue though.. Maybe it's a circular way that got simplified down to just it's start/end point.  I feel like I saw this recently in another recent issue.

**Comment by naoliv** [comment]
Sure, my description was incomplete, sorry.
What I was trying to say is "iD is creating highways with only 2 nodes, and they (initial and final) are the same".

**Comment by slhh** [comment]
The needle replacement in way.js will be a source of duplicated nodes if it is called for a way containing replacement as one of the neighbors of needle.

    replaceNode: function(needle, replacement) {
        if (this.nodes.indexOf(needle) < 0)
            return this;
        var nodes = this.nodes.slice();
        for (var i = 0; i < nodes.length; i++) {
            if (nodes[i] === needle) {
                nodes[i] = replacement;
            }
        }
        return this.update({nodes: nodes});
    },

I'm not sure if this can really happen. I will make a pull request to prevent this if desired.

**Comment by slhh** [comment]
The functions updateNode and addNode do have the same potential issue. The issue might be prevented from occuring by the calling code, but I think such basic functions should be more robust.

@bhousel I have a set of aditional tests and new implementations ready, but there is one existing test left.
For addNode there is a test for negativ Index counting from the end. Do we really want to have this behavior? It might simply be in the tests because it is the current behavior due to the splice function.

**Comment by bhousel** [comment]
Closing, because I think this was fixed in #3676 and #3750 
Please open a new issue if you see it in any version after iD 2.1
